In [ ]:
import pandas as pd
import numpy as np
import datetime as dt

import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
# 한글 폰트 사용을 위해서 세팅
from matplotlib import font_manager, rc
font_path = "C:/Windows/Fonts/NGULIM.TTF"
font = font_manager.FontProperties(fname=font_path).get_name()
rc('font', family=font)

In [ ]:
station = pd.read_csv('C:/Users/82105/Desktop/DNA와 함께 타슈/타슈/station_data.csv',encoding='cp949')

In [ ]:
station

번호                    정류장                    위치   Latitude   Longitude
0      1         무역전시관입구(택시승강장)           유성구 도룡동 3-8  36.374708  127.389027
1      2                대전컨벤션센터          유성구 도룡동 4-19  36.374901  127.391683
2      3                 한밭수목원1            서구 만년동 396  36.367485  127.388022
3      4      초원아파트(104동 버스정류장)            서구 만년동 401  36.367947  127.379519
4      5        둔산대공원 입구(버스정류장)       서구 둔산2동 1521-10  36.364505  127.388977
..   ...                    ...                   ...        ...         ...
297  338         크로바아파트(정문 맞은편)   서구 둔산1동 둔산1동 1554-1  36.351248  127.392028
298  339              청솔아파트(9동)       서구 둔산동 둔산동 2161  36.355551  127.389945
299  340                 오페라웨딩홀       서구 둔산동 둔산동 1257  36.348693  127.382585
300  341   대전종합고용지원센터(탄방역 4번출구)     서구 탄방동 탄방동 1453-1  36.346332  127.384968
301  342                 한가람아파트       서구 탄방동 탄방동 1447  36.347648  127.395212

[302 rows x 5 columns]

In [ ]:
ta_2021= pd.read_excel("C:/Users/82105/Desktop/DNA와 함께 타슈/타슈/2021타슈이용현황(전처리)_rbdus.xlsx")

C:\Users\82105\AppData\Local\Temp\ipykernel_18532\4208918240.py:1: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  ta_2021= pd.read_excel("C:/Users/82105/Desktop/DNA와 함께 타슈/타슈/2021타슈이용현황(전처리)_rbdus.xlsx")


In [ ]:
ta_2021

대여스테이션                대여일시        대여일  대여 위도  대여 경도  반납스테이션  \
0        135.0 2021-01-01 00:00:52 2021-01-01    NaN    NaN   139.0   
1         53.0 2021-01-01 05:13:50 2021-01-01    NaN    NaN    53.0   
2        154.0 2021-01-01 05:40:15 2021-01-01    NaN    NaN   132.0   
3        196.0 2021-01-01 06:26:01 2021-01-01    NaN    NaN   209.0   
4        149.0 2021-01-01 06:28:08 2021-01-01    NaN    NaN     NaN   
...        ...                 ...        ...    ...    ...     ...   
522711   317.0 2021-12-31 23:44:40 2021-12-31    NaN    NaN   327.0   
522712    37.0 2021-12-31 23:54:40 2021-12-31    NaN    NaN   203.0   
522713   151.0 2021-12-31 23:54:58 2021-12-31    NaN    NaN    82.0   
522714   139.0 2021-12-31 23:56:50 2021-12-31    NaN    NaN   139.0   
522715   139.0 2021-12-31 23:59:50 2021-12-31    NaN    NaN   139.0   

                       반납일시        반납일  반납 경도  반납 위도  이동거리(m)  회원구분  
0       2021-01-01 00:03:44 2021-01-01    NaN    NaN    290.0     1  
1       2021-01-01 05:36:19 2021-01-01    NaN    NaN   1750.0     2  
2       2021-01-01 09:49:33 2021-01-01    NaN    NaN   1750.0     2  
3       2021-01-01 06:34:04 2021-01-01    NaN    NaN    370.0     2  
4       2021-01-01 07:10:36 2021-01-01    NaN    NaN      NaN     0  
...                     ...        ...    ...    ...      ...   ...  
522711  2022-01-01 00:18:14 2022-01-01    NaN    NaN   2722.0     1  
522712  2022-01-01 00:37:04 2022-01-01    NaN    NaN   3880.0     1  
522713  2022-01-01 00:25:44 2022-01-01    NaN    NaN   2360.0     2  
522714  2021-12-31 23:57:08 2021-12-31    NaN    NaN      0.0     2  
522715  2022-01-01 00:00:12 2022-01-01    NaN    NaN      0.0     2  

[522716 rows x 12 columns]

In [ ]:
ta_2021.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 522716 entries, 0 to 522715
Data columns (total 12 columns):
 #   Column   Non-Null Count   Dtype         
---  ------   --------------   -----         
 0   대여스테이션   522676 non-null  float64       
 1   대여일시     522716 non-null  datetime64[ns]
 2   대여일      522716 non-null  datetime64[ns]
 3   대여 위도    0 non-null       float64       
 4   대여 경도    0 non-null       float64       
 5   반납스테이션   513602 non-null  float64       
 6   반납일시     522688 non-null  object        
 7   반납일      522688 non-null  datetime64[ns]
 8   반납 경도    0 non-null       float64       
 9   반납 위도    0 non-null       float64       
 10  이동거리(m)  513569 non-null  float64       
 11  회원구분     522716 non-null  int64         
dtypes: datetime64[ns](3), float64(7), int64(1), object(1)
memory usage: 47.9+ MB


## 결측값 확인

In [ ]:
ta_2021.isnull().sum()

대여스테이션         40
대여일시            0
대여일             0
대여 위도      522716
대여 경도      522716
반납스테이션       9114
반납일시           28
반납일            28
반납 경도      522716
반납 위도      522716
이동거리(m)      9147
회원구분            0
dtype: int64

* 대여스테이션 40개
* 반납스테이션 9114개
* 반납일시 28개

> 확인 필요

### 대여스테이션 결측갑 처리

In [ ]:
ta_2021[ta_2021['대여스테이션'].isnull()]

대여스테이션                대여일시        대여일  대여 위도  대여 경도  반납스테이션  \
418373     NaN 2021-10-07 20:34:40 2021-10-07    NaN    NaN     NaN   
419110     NaN 2021-10-08 09:29:57 2021-10-08    NaN    NaN   323.0   
423260     NaN 2021-10-10 07:52:26 2021-10-10    NaN    NaN     NaN   
423331     NaN 2021-10-10 09:14:49 2021-10-10    NaN    NaN   337.0   
423582     NaN 2021-10-10 11:51:07 2021-10-10    NaN    NaN     1.0   
423586     NaN 2021-10-10 11:51:56 2021-10-10    NaN    NaN     1.0   
433128     NaN 2021-10-16 15:13:02 2021-10-16    NaN    NaN    10.0   
434007     NaN 2021-10-17 12:42:26 2021-10-17    NaN    NaN     0.0   
436388     NaN 2021-10-18 20:29:46 2021-10-18    NaN    NaN     0.0   
438595     NaN 2021-10-20 16:02:52 2021-10-20    NaN    NaN     0.0   
438747     NaN 2021-10-20 17:10:04 2021-10-20    NaN    NaN   329.0   
439315     NaN 2021-10-20 21:22:21 2021-10-20    NaN    NaN     0.0   
439386     NaN 2021-10-20 22:28:23 2021-10-20    NaN    NaN     0.0   
439613     NaN 2021-10-21 08:25:35 2021-10-21    NaN    NaN   340.0   
439637     NaN 2021-10-21 08:35:26 2021-10-21    NaN    NaN     0.0   
440264     NaN 2021-10-21 16:32:15 2021-10-21    NaN    NaN   160.0   
445030     NaN 2021-10-24 11:19:29 2021-10-24    NaN    NaN     NaN   
445241     NaN 2021-10-24 13:09:52 2021-10-24    NaN    NaN     NaN   
446835     NaN 2021-10-25 08:37:12 2021-10-25    NaN    NaN   340.0   
447236     NaN 2021-10-25 15:40:10 2021-10-25    NaN    NaN   323.0   
448985     NaN 2021-10-26 15:43:44 2021-10-26    NaN    NaN   318.0   
450442     NaN 2021-10-27 10:42:59 2021-10-27    NaN    NaN   315.0   
453467     NaN 2021-10-28 20:10:29 2021-10-28    NaN    NaN     NaN   
454020     NaN 2021-10-29 08:39:43 2021-10-29    NaN    NaN   340.0   
454844     NaN 2021-10-29 17:15:57 2021-10-29    NaN    NaN     0.0   
455270     NaN 2021-10-29 19:28:35 2021-10-29    NaN    NaN     NaN   
455271     NaN 2021-10-29 19:28:48 2021-10-29    NaN    NaN     NaN   
456023     NaN 2021-10-30 11:45:30 2021-10-30    NaN    NaN   320.0   
456895     NaN 2021-10-30 17:28:50 2021-10-30    NaN    NaN   320.0   
459726     NaN 2021-11-01 08:37:06 2021-11-01    NaN    NaN   340.0   
460478     NaN 2021-11-01 17:12:24 2021-11-01    NaN    NaN     0.0   
465153     NaN 2021-11-03 22:18:56 2021-11-03    NaN    NaN     0.0   
467310     NaN 2021-11-05 08:40:04 2021-11-05    NaN    NaN   340.0   
477087     NaN 2021-11-12 21:00:54 2021-11-12    NaN    NaN     0.0   
477363     NaN 2021-11-13 10:18:13 2021-11-13    NaN    NaN     0.0   
490768     NaN 2021-11-22 12:56:21 2021-11-22    NaN    NaN   329.0   
501919     NaN 2021-12-04 15:12:50 2021-12-04    NaN    NaN     0.0   
502176     NaN 2021-12-04 20:50:10 2021-12-04    NaN    NaN     0.0   
509273     NaN 2021-12-12 14:49:26 2021-12-12    NaN    NaN     0.0   
521200     NaN 2021-12-29 08:35:49 2021-12-29    NaN    NaN   340.0   

                       반납일시        반납일  반납 경도  반납 위도  이동거리(m)  회원구분  
418373  2021-10-07 20:39:04 2021-10-07    NaN    NaN      2.0     1  
419110  2021-10-08 09:40:08 2021-10-08    NaN    NaN     13.0     0  
423260  2021-10-10 10:52:19 2021-10-10    NaN    NaN     13.0     0  
423331  2021-10-10 09:37:39 2021-10-10    NaN    NaN   1662.0     1  
423582  2021-10-10 12:21:40 2021-10-10    NaN    NaN   1865.0     1  
423586  2021-10-10 12:22:02 2021-10-10    NaN    NaN   1866.0     1  
433128  2021-10-16 15:29:40 2021-10-16    NaN    NaN    932.0     1  
434007  2021-10-17 13:01:14 2021-10-17    NaN    NaN   2285.0     0  
436388  2021-10-18 20:40:17 2021-10-18    NaN    NaN     14.0     0  
438595  2021-10-20 16:19:26 2021-10-20    NaN    NaN    949.0     1  
438747  2021-10-20 17:16:31 2021-10-20    NaN    NaN      9.0     0  
439315  2021-10-20 22:02:24 2021-10-20    NaN    NaN    875.0     1  
439386  2021-10-20 23:44:28 2021-10-20    NaN    NaN    784.0     1  
439613  2021-10-21 08:38:26 2021-10-21    NaN    NaN   1280.0     1  
439637  2021-10-21 08:45:39 2021-10-21    NaN 

- 대여스테이션이 없는 것들은 결측값 제거
- 이유: 대여스테이션이 아닌 곳에서 대여를 했을 수 있기 때문에 필요없다고 판단

#### 결측값 제거

In [ ]:
tasu_2021 = ta_2021[~ ta_2021['대여스테이션'].isnull()]

### 반납 스테이션 결측값 처리

In [ ]:
ta_2021[ta_2021['반납스테이션'].isnull()]

대여스테이션                대여일시        대여일  대여 위도  대여 경도  반납스테이션  \
4        149.0 2021-01-01 06:28:08 2021-01-01    NaN    NaN     NaN   
7        187.0 2021-01-01 07:37:53 2021-01-01    NaN    NaN     NaN   
115      188.0 2021-01-01 22:40:41 2021-01-01    NaN    NaN     NaN   
136      200.0 2021-01-02 11:44:49 2021-01-02    NaN    NaN     NaN   
303      147.0 2021-01-02 19:34:56 2021-01-02    NaN    NaN     NaN   
...        ...                 ...        ...    ...    ...     ...   
522659   179.0 2021-12-31 20:04:46 2021-12-31    NaN    NaN     NaN   
522663   304.0 2021-12-31 20:16:39 2021-12-31    NaN    NaN     NaN   
522676    29.0 2021-12-31 20:58:55 2021-12-31    NaN    NaN     NaN   
522702    66.0 2021-12-31 22:57:16 2021-12-31    NaN    NaN     NaN   
522709   186.0 2021-12-31 23:41:22 2021-12-31    NaN    NaN     NaN   

                       반납일시        반납일  반납 경도  반납 위도  이동거리(m)  회원구분  
4       2021-01-01 07:10:36 2021-01-01    NaN    NaN      NaN     0  
7       2021-01-01 07:53:01 2021-01-01    NaN    NaN      NaN     0  
115     2021-01-02 00:12:06 2021-01-02    NaN    NaN      NaN     2  
136     2021-01-02 12:20:17 2021-01-02    NaN    NaN      NaN     0  
303     2021-01-02 22:51:54 2021-01-02    NaN    NaN      NaN     2  
...                     ...        ...    ...    ...      ...   ...  
522659  2021-12-31 20:52:23 2021-12-31    NaN    NaN      NaN     0  
522663  2021-12-31 22:20:46 2021-12-31    NaN    NaN      NaN     1  
522676  2021-12-31 23:07:08 2021-12-31    NaN    NaN      NaN     1  
522702  2021-12-31 23:08:23 2021-12-31    NaN    NaN      NaN     1  
522709  2022-01-01 17:08:11 2022-01-01    NaN    NaN      NaN     2  

[9114 rows x 12 columns]

* 반납스테이션도 대여스테이션과 동일한 이유로 제거

#### 결측값 제거

In [ ]:
tasu_2021 = tasu_2021[~ tasu_2021['반납스테이션'].isnull()]

In [ ]:
tasu_2021.isnull().sum()

대여스테이션          0
대여일시            0
대여일             0
대여 위도      513569
대여 경도      513569
반납스테이션          0
반납일시            0
반납일             0
반납 경도      513569
반납 위도      513569
이동거리(m)      1646
회원구분            0
dtype: int64

### 반납일시 결측값 처리

In [ ]:
ta_2021[ta_2021['반납일시'].isnull()]

대여스테이션                대여일시        대여일  대여 위도  대여 경도  반납스테이션 반납일시 반납일  \
210102    14.0 2021-06-08 17:51:26 2021-06-08    NaN    NaN     NaN  NaN NaT   
211744   144.0 2021-06-09 12:01:37 2021-06-09    NaN    NaN     NaN  NaN NaT   
211759   144.0 2021-06-09 12:11:22 2021-06-09    NaN    NaN     NaN  NaN NaT   
211917   144.0 2021-06-09 15:23:45 2021-06-09    NaN    NaN     NaN  NaN NaT   
212019   144.0 2021-06-09 16:39:15 2021-06-09    NaN    NaN     NaN  NaN NaT   
222962   144.0 2021-06-16 16:03:45 2021-06-16    NaN    NaN     NaN  NaN NaT   
222988   144.0 2021-06-16 16:17:43 2021-06-16    NaN    NaN     NaN  NaN NaT   
223039   144.0 2021-06-16 16:45:40 2021-06-16    NaN    NaN     NaN  NaN NaT   
223069   144.0 2021-06-16 16:57:12 2021-06-16    NaN    NaN     NaN  NaN NaT   
223122   144.0 2021-06-16 17:21:24 2021-06-16    NaN    NaN     NaN  NaN NaT   
268969     0.0 2021-07-16 23:28:32 2021-07-16    NaN    NaN     NaN  NaN NaT   
268971     0.0 2021-07-16 23:31:14 2021-07-16    NaN    NaN     NaN  NaN NaT   
270662   329.0 2021-07-18 08:23:29 2021-07-18    NaN    NaN     NaN  NaN NaT   
270664   329.0 2021-07-18 08:24:08 2021-07-18    NaN    NaN     NaN  NaN NaT   
270665   329.0 2021-07-18 08:24:45 2021-07-18    NaN    NaN     NaN  NaN NaT   
270666   329.0 2021-07-18 08:25:39 2021-07-18    NaN    NaN     NaN  NaN NaT   
270669   329.0 2021-07-18 08:29:42 2021-07-18    NaN    NaN     NaN  NaN NaT   
272141   329.0 2021-07-19 17:32:48 2021-07-19    NaN    NaN     NaN  NaN NaT   
272143   329.0 2021-07-19 17:33:24 2021-07-19    NaN    NaN     NaN  NaN NaT   
272688   323.0 2021-07-19 22:15:54 2021-07-19    NaN    NaN     NaN  NaN NaT   
272690   323.0 2021-07-19 22:16:40 2021-07-19    NaN    NaN     NaN  NaN NaT   
272921   339.0 2021-07-20 06:49:50 2021-07-20    NaN    NaN     NaN  NaN NaT   
272922   339.0 2021-07-20 06:50:27 2021-07-20    NaN    NaN     NaN  NaN NaT   
273143   323.0 2021-07-20 09:27:42 2021-07-20    NaN    NaN     NaN  NaN NaT   
273144   323.0 2021-07-20 09:28:36 2021-07-20    NaN    NaN     NaN  NaN NaT   
273147   323.0 2021-07-20 09:30:42 2021-07-20    NaN    NaN     NaN  NaN NaT   
273151   323.0 2021-07-20 09:32:27 2021-07-20    NaN    NaN     NaN  NaN NaT   
289060   332.0 2021-07-30 13:05:54 2021-07-30    NaN    NaN     NaN  NaN NaT   

        반납 경도  반납 위도  이동거리(m)  회원구분  
210102    NaN    NaN      NaN     1  
211744    NaN    NaN      NaN     0  
211759    NaN    NaN      NaN     0  
211917    NaN    NaN      NaN     0  
212019    NaN    NaN      NaN     0  
222962    NaN    NaN      NaN     1  
222988    NaN    NaN      NaN     0  
223039    NaN    NaN      NaN     0  
223069    NaN    NaN      NaN     0  
223122    NaN    NaN      NaN     0  
268969    NaN    NaN      NaN     1  
268971    NaN    NaN      NaN     1  
270662    NaN    NaN      NaN     0  
270664    NaN    NaN      NaN     0  
270665    NaN    NaN      NaN     0  
270666    NaN    NaN      NaN     0  
270669    NaN    NaN      NaN     0  
272141    NaN    NaN      NaN     0  
272143    NaN    NaN      NaN     0  
272688    NaN    NaN      NaN     0  
272690    NaN    NaN      NaN     0  
272921    NaN    NaN      NaN     0  
272922    NaN    NaN      NaN     0  
273143    NaN    NaN      NaN     0  
273144    NaN    NaN      NaN     0  
273147    NaN    NaN      NaN     0  
273151    NaN    NaN      NaN     0  
289060    NaN    NaN      NaN     0

- 반납일시 결측값 데이터 제거
- 이유: 반납일시가 없는 데이터들을 보면 대여스테이션만 나와있고, 반납스테이션은 안나와 있음


#### 결측값 제거

In [ ]:
tasu_2021 = tasu_2021[~ tasu_2021['반납일시'].isnull()]

In [ ]:
tasu_2021.isnull().sum()

대여스테이션          0
대여일시            0
대여일             0
대여 위도      513569
대여 경도      513569
반납스테이션          0
반납일시            0
반납일             0
반납 경도      513569
반납 위도      513569
이동거리(m)      1646
회원구분            0
dtype: int64

## 데이터 타입 변경
1. 대여 스테이션 >> int
2. 반납 스테이션 >> int

In [ ]:
tasu_2021.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 513569 entries, 0 to 522715
Data columns (total 12 columns):
 #   Column   Non-Null Count   Dtype         
---  ------   --------------   -----         
 0   대여스테이션   513569 non-null  float64       
 1   대여일시     513569 non-null  datetime64[ns]
 2   대여일      513569 non-null  datetime64[ns]
 3   대여 위도    0 non-null       float64       
 4   대여 경도    0 non-null       float64       
 5   반납스테이션   513569 non-null  float64       
 6   반납일시     513569 non-null  object        
 7   반납일      513569 non-null  datetime64[ns]
 8   반납 경도    0 non-null       float64       
 9   반납 위도    0 non-null       float64       
 10  이동거리(m)  511923 non-null  float64       
 11  회원구분     513569 non-null  int64         
dtypes: datetime64[ns](3), float64(7), int64(1), object(1)
memory usage: 50.9+ MB


In [ ]:
tasu_2021 = tasu_2021.astype({'대여스테이션':'int'})
tasu_2021 = tasu_2021.astype({'반납스테이션':'int'})

### station와 tasu_2021의 대여서 번호값 확인

In [ ]:
tasu_2021['대여스테이션'].value_counts()

0      21436
3      17212
186    16468
31      8128
17      7031
       ...  
306      120
128      118
211      103
301       53
302       44
Name: 대여스테이션, Length: 304, dtype: int64

In [ ]:
station['번호'].value_counts()

1      1
210    1
208    1
207    1
206    1
      ..
102    1
101    1
100    1
99     1
342    1
Name: 번호, Length: 302, dtype: int64

## 대여스테이션, 반납스테이션 0 값인 행 제거

- 이유: 0은 대여소가 아닌 곳이라고 판단

In [ ]:
tasu_2021.drop(tasu_2021.loc[tasu_2021['대여스테이션']== 0].index, inplace = True)

In [ ]:
tasu_2021.drop(tasu_2021.loc[tasu_2021['반납스테이션']== 0].index, inplace = True)

In [ ]:
station1=station.rename(columns = {'번호':'대여스테이션'})

In [ ]:
total1 = pd.merge(tasu_2021, station1, how = 'outer', on='대여스테이션')

In [ ]:
station2=station1.rename(columns = {'대여스테이션':'반납스테이션'})

In [ ]:
total = pd.merge(total1, station2, how = 'outer', on= '반납스테이션')

In [ ]:
total

대여스테이션                대여일시        대여일  대여 위도  대여 경도  반납스테이션  \
0          135 2021-01-01 00:00:52 2021-01-01    NaN    NaN     139   
1          135 2021-01-16 21:43:10 2021-01-16    NaN    NaN     139   
2          135 2021-02-22 18:51:19 2021-02-22    NaN    NaN     139   
3          135 2021-02-24 19:01:10 2021-02-24    NaN    NaN     139   
4          135 2021-02-27 22:08:14 2021-02-27    NaN    NaN     139   
...        ...                 ...        ...    ...    ...     ...   
485940     329 2021-10-28 05:35:52 2021-10-28    NaN    NaN     332   
485941     329 2021-10-29 05:36:15 2021-10-29    NaN    NaN     332   
485942     325 2021-07-28 21:36:38 2021-07-28    NaN    NaN     332   
485943     325 2021-08-08 22:50:51 2021-08-08    NaN    NaN     332   
485944     325 2021-10-09 17:39:31 2021-10-09    NaN    NaN     332   

                       반납일시        반납일  반납 경도  반납 위도  이동거리(m)  회원구분  \
0       2021-01-01 00:03:44 2021-01-01    NaN    NaN    290.0     1   
1       2021-01-16 21:46:26 2021-01-16    NaN    NaN    310.0     0   
2       2021-02-22 18:56:30 2021-02-22    NaN    NaN    440.0     1   
3       2021-02-24 19:06:27 2021-02-24    NaN    NaN    360.0     0   
4       2021-02-27 22:31:58 2021-02-27    NaN    NaN   1230.0     1   
...                     ...        ...    ...    ...      ...   ...   
485940  2021-10-28 05:48:10 2021-10-28    NaN    NaN   1636.0     1   
485941  2021-10-29 05:49:10 2021-10-29    NaN    NaN   1336.0     1   
485942  2021-07-28 23:28:28 2021-07-28    NaN    NaN    755.0     1   
485943  2021-08-08 22:59:55 2021-08-08    NaN    NaN    748.0     0   
485944  2021-10-09 17:52:08 2021-10-09    NaN    NaN    730.0     1   

                정류장_x                위치_x  Latitude_x  Longitude_x  \
0             서대전역네거리        중구 오류동 168-7   36.320935   127.406069   
1             서대전역네거리        중구 오류동 168-7   36.320935   127.406069   
2             서대전역네거리        중구 오류동 168-7   36.320935   127.406069   
3             서대전역네거리        중구 오류동 168-7   36.320935   127.406069   
4             서대전역네거리        중구 오류동 168-7   36.320935   127.406069   
...               ...                 ...         ...          ...   
485940       갈마역 2번출구   서구 둔산동 둔산동 1515-2   36.357586   127.375422   
485941       갈마역 2번출구   서구 둔산동 둔산동 1515-2   36.357586   127.375422   
485942   수정타운아파트(13동)     서구 둔산동 둔산동 1515   36.357606   127.395179   
485943   수정타운아파트(13동)     서구 둔산동 둔산동 1515   36.357606   127.395179   
485944   수정타운아파트(13동)     서구 둔산동 둔산동 1515   36.357606   127.395179   

                     정류장_y                  위치_y  Latitude_y  Longitude_y  
0             서대전네거리역 3번출구          중구 오류동 152-5   36.322330   127.411393  
1             서대전네거리역 3번출구          중구 오류동 152-5   36.322330   127.411393  
2             서대전네거리역 3번출구          중구 오류동 152-5   36.322330   127.411393  
3             서대전네거리역 3번출구          중구 오류동 152-5   36.322330   127.411393  
4             서대전네거리역 3번출구          중구 오류동 152-5   36.322330   127.411393  
...                    ...                   ...         ...          ...  
485940   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   36.356279   127.390021  
485941   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   36.356279   127.390021  
485942   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   36.356279   127.390021  
485943   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   36.356279   127.390021  
485944   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   36.356279   127.390021  

[485945 rows x 20 columns]

In [ ]:
total=total.drop(['대여 위도'], axis = 1)
total=total.drop(['대여 경도'], axis = 1)
total=total.drop(['반납 위도'], axis = 1)
total=total.drop(['반납 경도'], axis = 1)

* 대여 반납 위도 경도 삭제

In [ ]:
total=total.rename(columns = {'정류장_x':'대여장소'})
total=total.rename(columns = {'위치_x':'대여위치'})
total=total.rename(columns = {'정류장_y':'반납장소'})
total=total.rename(columns = {'위치_y':'반납위치'})
total=total.rename(columns = {'Latitude_x':'대여위도'})
total=total.rename(columns = {'Longitude_x':'대여경도'})
total=total.rename(columns = {'Latitude_y':'반납위도'})
total=total.rename(columns = {'Longitude_y':'반납경도'})


In [ ]:
* 컬럼명 변경

SyntaxError: invalid syntax (1385492610.py, line 1)

In [ ]:
total

대여스테이션                대여일시        대여일  반납스테이션                 반납일시  \
0          135 2021-01-01 00:00:52 2021-01-01     139  2021-01-01 00:03:44   
1          135 2021-01-16 21:43:10 2021-01-16     139  2021-01-16 21:46:26   
2          135 2021-02-22 18:51:19 2021-02-22     139  2021-02-22 18:56:30   
3          135 2021-02-24 19:01:10 2021-02-24     139  2021-02-24 19:06:27   
4          135 2021-02-27 22:08:14 2021-02-27     139  2021-02-27 22:31:58   
...        ...                 ...        ...     ...                  ...   
485940     329 2021-10-28 05:35:52 2021-10-28     332  2021-10-28 05:48:10   
485941     329 2021-10-29 05:36:15 2021-10-29     332  2021-10-29 05:49:10   
485942     325 2021-07-28 21:36:38 2021-07-28     332  2021-07-28 23:28:28   
485943     325 2021-08-08 22:50:51 2021-08-08     332  2021-08-08 22:59:55   
485944     325 2021-10-09 17:39:31 2021-10-09     332  2021-10-09 17:52:08   

              반납일  이동거리(m)  회원구분           대여장소                대여위치  \
0      2021-01-01    290.0     1        서대전역네거리        중구 오류동 168-7   
1      2021-01-16    310.0     0        서대전역네거리        중구 오류동 168-7   
2      2021-02-22    440.0     1        서대전역네거리        중구 오류동 168-7   
3      2021-02-24    360.0     0        서대전역네거리        중구 오류동 168-7   
4      2021-02-27   1230.0     1        서대전역네거리        중구 오류동 168-7   
...           ...      ...   ...            ...                 ...   
485940 2021-10-28   1636.0     1       갈마역 2번출구   서구 둔산동 둔산동 1515-2   
485941 2021-10-29   1336.0     1       갈마역 2번출구   서구 둔산동 둔산동 1515-2   
485942 2021-07-28    755.0     1   수정타운아파트(13동)     서구 둔산동 둔산동 1515   
485943 2021-08-08    748.0     0   수정타운아파트(13동)     서구 둔산동 둔산동 1515   
485944 2021-10-09    730.0     1   수정타운아파트(13동)     서구 둔산동 둔산동 1515   

             대여위도        대여경도                반납장소                  반납위치  \
0       36.320935  127.406069        서대전네거리역 3번출구          중구 오류동 152-5   
1       36.320935  127.406069        서대전네거리역 3번출구          중구 오류동 152-5   
2       36.320935  127.406069        서대전네거리역 3번출구          중구 오류동 152-5   
3       36.320935  127.406069        서대전네거리역 3번출구          중구 오류동 152-5   
4       36.320935  127.406069        서대전네거리역 3번출구          중구 오류동 152-5   
...           ...         ...                 ...                   ...   
485940  36.357586  127.375422   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   
485941  36.357586  127.375422   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   
485942  36.357606  127.395179   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   
485943  36.357606  127.395179   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   
485944  36.357606  127.395179   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   

             반납위도        반납경도  
0       36.322330  127.411393  
1       36.322330  127.411393  
2       36.322330  127.411393  
3       36.322330  127.411393  
4       36.322330  127.411393  
...           ...         ...  
485940  36.356279  127.390021  
485941  36.356279  127.390021  
485942  36.356279  127.390021  
485943  36.356279  127.390021  
485944  36.356279  127.390021  

[485945 rows x 16 columns]

In [ ]:
total[total['대여위치'].isnull()]

대여스테이션                대여일시        대여일  반납스테이션                 반납일시  \
355        105 2021-04-24 13:28:08 2021-04-24     139  2021-04-24 14:25:21   
5886       105 2021-05-29 15:14:40 2021-05-29      51  2021-05-29 16:08:53   
34041      105 2021-05-09 07:57:38 2021-05-09      93  2021-05-09 08:58:32   
51800      105 2021-02-06 11:18:32 2021-02-06      17  2021-02-06 11:46:14   
51801      105 2021-03-13 14:03:22 2021-03-13      17  2021-03-13 14:25:16   
...        ...                 ...        ...     ...                  ...   
484923     105 2021-12-20 23:19:35 2021-12-20     304  2021-12-20 23:36:16   
484924     105 2021-12-24 18:40:05 2021-12-24     304  2021-12-24 18:57:33   
484925     105 2021-12-27 23:46:24 2021-12-27     304  2021-12-28 00:10:45   
484926     105 2021-12-29 22:27:57 2021-12-29     304  2021-12-29 22:43:48   
485521     105 2021-11-18 17:46:10 2021-11-18     307  2021-11-18 18:02:58   

              반납일  이동거리(m)  회원구분 대여장소 대여위치  대여위도  대여경도            반납장소  \
355    2021-04-24   5320.0     0  NaN  NaN   NaN   NaN    서대전네거리역 3번출구   
5886   2021-05-29   3910.0     0  NaN  NaN   NaN   NaN       남선공원종합체육관   
34041  2021-05-09    720.0     2  NaN  NaN   NaN   NaN           서대전육교   
51800  2021-02-06   2220.0     2  NaN  NaN   NaN   NaN            타임월드   
51801  2021-03-13   2040.0     2  NaN  NaN   NaN   NaN            타임월드   
...           ...      ...   ...  ...  ...   ...   ...             ...   
484923 2021-12-20   1674.0     0  NaN  NaN   NaN   NaN        충대 서문네거리   
484924 2021-12-24   1836.0     0  NaN  NaN   NaN   NaN        충대 서문네거리   
484925 2021-12-28   1848.0     0  NaN  NaN   NaN   NaN        충대 서문네거리   
484926 2021-12-29   1846.0     0  NaN  NaN   NaN   NaN        충대 서문네거리   
485521 2021-11-18   1339.0     0  NaN  NaN   NaN   NaN   충대정문 오거리1(정문)   

                     반납위치       반납위도        반납경도  
355          중구 오류동 152-5  36.322330  127.411393  
5886          서구 탄방동 1458  36.348872  127.398631  
34041        중구 유천동 186-7  36.319003  127.400926  
51800        서구 둔산2동 1036  36.352308  127.378549  
51801        서구 둔산2동 1036  36.352308  127.378549  
...                   ...        ...         ...  
484923   유성구 죽동 죽동 725-11  36.377061  127.332834  
484924   유성구 죽동 죽동 725-11  36.377061  127.332834  
484925   유성구 죽동 죽동 725-11  36.377061  127.332834  
484926   유성구 죽동 죽동 725-11  36.377061  127.332834  
485521    유성구 궁동 궁동 258-8  36.362512  127.344037  

[3030 rows x 16 columns]

In [ ]:
total[total['대여장소'].isnull()]

대여스테이션                대여일시        대여일  반납스테이션                 반납일시  \
355        105 2021-04-24 13:28:08 2021-04-24     139  2021-04-24 14:25:21   
5886       105 2021-05-29 15:14:40 2021-05-29      51  2021-05-29 16:08:53   
34041      105 2021-05-09 07:57:38 2021-05-09      93  2021-05-09 08:58:32   
51800      105 2021-02-06 11:18:32 2021-02-06      17  2021-02-06 11:46:14   
51801      105 2021-03-13 14:03:22 2021-03-13      17  2021-03-13 14:25:16   
...        ...                 ...        ...     ...                  ...   
484923     105 2021-12-20 23:19:35 2021-12-20     304  2021-12-20 23:36:16   
484924     105 2021-12-24 18:40:05 2021-12-24     304  2021-12-24 18:57:33   
484925     105 2021-12-27 23:46:24 2021-12-27     304  2021-12-28 00:10:45   
484926     105 2021-12-29 22:27:57 2021-12-29     304  2021-12-29 22:43:48   
485521     105 2021-11-18 17:46:10 2021-11-18     307  2021-11-18 18:02:58   

              반납일  이동거리(m)  회원구분 대여장소 대여위치  대여위도  대여경도            반납장소  \
355    2021-04-24   5320.0     0  NaN  NaN   NaN   NaN    서대전네거리역 3번출구   
5886   2021-05-29   3910.0     0  NaN  NaN   NaN   NaN       남선공원종합체육관   
34041  2021-05-09    720.0     2  NaN  NaN   NaN   NaN           서대전육교   
51800  2021-02-06   2220.0     2  NaN  NaN   NaN   NaN            타임월드   
51801  2021-03-13   2040.0     2  NaN  NaN   NaN   NaN            타임월드   
...           ...      ...   ...  ...  ...   ...   ...             ...   
484923 2021-12-20   1674.0     0  NaN  NaN   NaN   NaN        충대 서문네거리   
484924 2021-12-24   1836.0     0  NaN  NaN   NaN   NaN        충대 서문네거리   
484925 2021-12-28   1848.0     0  NaN  NaN   NaN   NaN        충대 서문네거리   
484926 2021-12-29   1846.0     0  NaN  NaN   NaN   NaN        충대 서문네거리   
485521 2021-11-18   1339.0     0  NaN  NaN   NaN   NaN   충대정문 오거리1(정문)   

                     반납위치       반납위도        반납경도  
355          중구 오류동 152-5  36.322330  127.411393  
5886          서구 탄방동 1458  36.348872  127.398631  
34041        중구 유천동 186-7  36.319003  127.400926  
51800        서구 둔산2동 1036  36.352308  127.378549  
51801        서구 둔산2동 1036  36.352308  127.378549  
...                   ...        ...         ...  
484923   유성구 죽동 죽동 725-11  36.377061  127.332834  
484924   유성구 죽동 죽동 725-11  36.377061  127.332834  
484925   유성구 죽동 죽동 725-11  36.377061  127.332834  
484926   유성구 죽동 죽동 725-11  36.377061  127.332834  
485521    유성구 궁동 궁동 258-8  36.362512  127.344037  

[3030 rows x 16 columns]

In [ ]:
total[total['반납장소'].isnull()]

대여스테이션                대여일시        대여일  반납스테이션                 반납일시  \
429170      38 2021-01-22 16:25:10 2021-01-22     105  2021-01-22 16:38:28   
429171      38 2021-03-05 09:32:43 2021-03-05     105  2021-03-05 09:42:34   
429172      38 2021-03-13 15:22:26 2021-03-13     105  2021-03-13 15:34:40   
429173      38 2021-04-05 09:29:10 2021-04-05     105  2021-04-05 09:39:16   
429174      38 2021-04-06 09:57:25 2021-04-06     105  2021-04-06 10:06:14   
...        ...                 ...        ...     ...                  ...   
433422     326 2021-10-13 10:11:45 2021-10-13     105  2021-10-13 10:19:08   
433423     326 2021-10-14 09:58:45 2021-10-14     105  2021-10-14 10:08:28   
433424     326 2021-10-27 08:40:24 2021-10-27     105  2021-10-27 08:53:29   
433425     326 2021-11-16 09:51:12 2021-11-16     105  2021-11-16 10:04:21   
433426     304 2021-12-14 09:42:16 2021-12-14     105  2021-12-14 09:58:03   

              반납일  이동거리(m)  회원구분       대여장소               대여위치       대여위도  \
429170 2021-01-22   1050.0     1   갑천역 1번출구        서구 월평동 1534  36.354360   
429171 2021-03-05   1100.0     1   갑천역 1번출구        서구 월평동 1534  36.354360   
429172 2021-03-13   1120.0     1   갑천역 1번출구        서구 월평동 1534  36.354360   
429173 2021-04-05   1040.0     1   갑천역 1번출구        서구 월평동 1534  36.354360   
429174 2021-04-06    960.0     1   갑천역 1번출구        서구 월평동 1534  36.354360   
...           ...      ...   ...        ...                ...        ...   
433422 2021-10-13      5.0     0   갑천역 1번출구    서구 월평동 월평동 1534  36.354360   
433423 2021-10-14      1.0     0   갑천역 1번출구    서구 월평동 월평동 1534  36.354360   
433424 2021-10-27   1059.0     1   갑천역 1번출구    서구 월평동 월평동 1534  36.354360   
433425 2021-11-16   1211.0     0   갑천역 1번출구    서구 월평동 월평동 1534  36.354360   
433426 2021-12-14   1847.0     0   충대 서문네거리   유성구 죽동 죽동 725-11  36.377061   

              대여경도 반납장소 반납위치  반납위도  반납경도  
429170  127.354601  NaN  NaN   NaN   NaN  
429171  127.354601  NaN  NaN   NaN   NaN  
429172  127.354601  NaN  NaN   NaN   NaN  
429173  127.354601  NaN  NaN   NaN   NaN  
429174  127.354601  NaN  NaN   NaN   NaN  
...            ...  ...  ...   ...   ...  
433422  127.354601  NaN  NaN   NaN   NaN  
433423  127.354601  NaN  NaN   NaN   NaN  
433424  127.354601  NaN  NaN   NaN   NaN  
433425  127.354601  NaN  NaN   NaN   NaN  
433426  127.332834  NaN  NaN   NaN   NaN  

[4257 rows x 16 columns]

타슈 이용 현황 데이터에서 105번이 없음

-> 결측값으로 처리해서 제거

In [ ]:
total = total[~ total['반납장소'].isnull()]
total = total[~ total['반납위치'].isnull()]
total = total[~ total['대여장소'].isnull()]
total = total[~ total['대여위치'].isnull()]

In [ ]:
total.isnull().sum()

대여스테이션       0
대여일시         0
대여일          0
반납스테이션       0
반납일시         0
반납일          0
이동거리(m)    688
회원구분         0
대여장소         0
대여위치         0
대여위도         0
대여경도         0
반납장소         0
반납위치         0
반납위도         0
반납경도         0
dtype: int64

결측값 처리 성공!!

## 데이터 저장

In [ ]:
total.to_csv('total.csv')

In [ ]:
total

대여스테이션                대여일시        대여일  반납스테이션                 반납일시  \
0          135 2021-01-01 00:00:52 2021-01-01     139  2021-01-01 00:03:44   
1          135 2021-01-16 21:43:10 2021-01-16     139  2021-01-16 21:46:26   
2          135 2021-02-22 18:51:19 2021-02-22     139  2021-02-22 18:56:30   
3          135 2021-02-24 19:01:10 2021-02-24     139  2021-02-24 19:06:27   
4          135 2021-02-27 22:08:14 2021-02-27     139  2021-02-27 22:31:58   
...        ...                 ...        ...     ...                  ...   
485940     329 2021-10-28 05:35:52 2021-10-28     332  2021-10-28 05:48:10   
485941     329 2021-10-29 05:36:15 2021-10-29     332  2021-10-29 05:49:10   
485942     325 2021-07-28 21:36:38 2021-07-28     332  2021-07-28 23:28:28   
485943     325 2021-08-08 22:50:51 2021-08-08     332  2021-08-08 22:59:55   
485944     325 2021-10-09 17:39:31 2021-10-09     332  2021-10-09 17:52:08   

              반납일  이동거리(m)  회원구분           대여장소                대여위치  \
0      2021-01-01    290.0     1        서대전역네거리        중구 오류동 168-7   
1      2021-01-16    310.0     0        서대전역네거리        중구 오류동 168-7   
2      2021-02-22    440.0     1        서대전역네거리        중구 오류동 168-7   
3      2021-02-24    360.0     0        서대전역네거리        중구 오류동 168-7   
4      2021-02-27   1230.0     1        서대전역네거리        중구 오류동 168-7   
...           ...      ...   ...            ...                 ...   
485940 2021-10-28   1636.0     1       갈마역 2번출구   서구 둔산동 둔산동 1515-2   
485941 2021-10-29   1336.0     1       갈마역 2번출구   서구 둔산동 둔산동 1515-2   
485942 2021-07-28    755.0     1   수정타운아파트(13동)     서구 둔산동 둔산동 1515   
485943 2021-08-08    748.0     0   수정타운아파트(13동)     서구 둔산동 둔산동 1515   
485944 2021-10-09    730.0     1   수정타운아파트(13동)     서구 둔산동 둔산동 1515   

             대여위도        대여경도                반납장소                  반납위치  \
0       36.320935  127.406069        서대전네거리역 3번출구          중구 오류동 152-5   
1       36.320935  127.406069        서대전네거리역 3번출구          중구 오류동 152-5   
2       36.320935  127.406069        서대전네거리역 3번출구          중구 오류동 152-5   
3       36.320935  127.406069        서대전네거리역 3번출구          중구 오류동 152-5   
4       36.320935  127.406069        서대전네거리역 3번출구          중구 오류동 152-5   
...           ...         ...                 ...                   ...   
485940  36.357586  127.375422   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   
485941  36.357586  127.375422   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   
485942  36.357606  127.395179   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   
485943  36.357606  127.395179   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   
485944  36.357606  127.395179   둔산 홈플러스(법원 버스정류장)   서구 둔산1동 둔산1동 1380-5   

             반납위도        반납경도  
0       36.322330  127.411393  
1       36.322330  127.411393  
2       36.322330  127.411393  
3       36.322330  127.411393  
4       36.322330  127.411393  
...           ...         ...  
485940  36.356279  127.390021  
485941  36.356279  127.390021  
485942  36.356279  127.390021  
485943  36.356279  127.390021  
485944  36.356279  127.390021  

[479216 rows x 16 columns]